MetaMetrics: Fixes TypeError: Object.values(...).reduce(...).flat is not a function

digiwand opened the pull request:
## Explanation

Sentry Issue: [METAMASK-P61Y](https://sentry.io/organizations/metamask/issues/3349418490/?referrer=github_integration)
```
TypeError: Object.values(...).reduce(...).flat is not a function
  at apply (app/scripts/controllers/metametrics.js:627:8)
  at _getAllUniqueNFTAddressesLength (app/scripts/controllers/metametrics.js:638:34)
  at _buildUserTraitsObject (app/scripts/controllers/metametrics.js:564:48)
  at handleMetaMaskStateUpdate (app/scripts/controllers/metametrics.js:469:29)
  at call (app/scripts/metamask-controller.js:330:7)
```

This issue is happening because `flat` is not support in some browsers or older versions of browsers. 

Ref:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat#browser_compatibility 
- https://caniuse.com/mdn-javascript_builtins_array_flat

## More Information

* Fixes https://github.com/MetaMask/metamask-extension/issues/14960
* Fixes 10.15.0

## Screenshots/Screencaps

<img width="1302" alt="Screenshot 2022-06-15 at 12 15 20 PM" src="https://user-images.githubusercontent.com/20778143/173886786-68f2bf5a-8d63-4259-b9f7-75f33345f0a0.png">


## Manual Testing Steps

TLDR; Test MetaMetrics feature using Internet Explorer

1. Turn on MetaMetrics
  a. Go to Settings > Security & Privacy
  b. Turn on Participate in MetaMetrics
  c. Ensure no ad blockers are on if you want to send events to Segment
2. Add NFTs to MM
  a. Faucet: https://faucet.paradigm.xyz/
  b. Use "Import Tokens" feature to add tokens by contract and token IDs
3. Toggle a setting to trigger the MetaMetricsController#_identify call that calls this errored function
4. Verify no error messages found from background.js when using internet explorer or another incompatible browser
5. Verify NFT user traits are being sent to Segment as expected


## Pre-Merge Checklist

- [x] PR template is filled out
- [x] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added (Test exists)
- [x] PR is linked to the appropriate GitHub issue
- [x] PR has been added to the appropriate release Milestone

### + If there are functional changes:

- [ ] Manual testing complete & passed
- [ ] "Extension QA Board" label has been applied

digiwand committed:
metametrics: flat is not a function

darkwing commented on app/scripts/controllers/metametrics.js:
```diff
@@ -630,11 +630,15 @@ export default class MetaMetricsController {
+    return Object.values(allCollectibles).reduce((result, chainNFTs) => {
+      let updatedResult;
+
+      Object.values(chainNFTs).forEach((addressNFTs) => {
```
I don't actually think we need the `forEach` here.  I think we can simply do:
```
return result.concat(Object.values(chainNFTs))
```

digiwand replied:
hm... I don't think this works

It might help to see this mock data I created:
```
const allCollectibles = {
          '0xac706cE8A9BF27Afecf080fB298d0ee13cfb9123': {
            56: [
              {
                address: '0xd2cea331e5f5d8ee9fb1055c297795937645de91',
                tokenId: '100',
              },
              {
                address: '0xd2cea331e5f5d8ee9fb1055c297795937645de91',
                tokenId: '101',
              },
              {
                address: '0x7488d2ce5deb26db021285b50b661d655eb3d3d9',
                tokenId: '99',
              },
            ],
          },
          '0xe04AB39684A24D8D4124b114F3bd6FBEB779c345: {
            69: [
              {
                address: '0x63d646bc7380562376d5de205123a57b1718184d',
                tokenId: '14',
              },
            ],
          },
};
```
 
replacing the logic with `return result.concat(Object.values(chainNFTs))` will produce this result
   
<img width="528" alt="Screenshot 2022-06-15 at 12 57 53 PM" src="https://user-images.githubusercontent.com/20778143/173894082-5101ee2a-8d69-484b-863e-34e94f04c52e.png">

darkwing replied:
I was wrong -- I meant to say `return result.concat(…Object.values(chainNFTs))`

digiwand replied:
You're totally right! I thought I was overengineering this when I wrote this out. Thank you @darkwing 🙏

digiwand replied:
<img width="420" alt="Screenshot 2022-06-15 at 12 15 20 PM" src="https://user-images.githubusercontent.com/20778143/173896535-af49b18e-6017-4bea-b929-d4ded025b06d.jpeg">

digiwand committed:
MetaMetrics: shorten array logic

digiwand commented:
I was thinking that I could have sworn I fixed this, and then I realized I removed `flatMap` and left `flat` in when I worked on this PR https://github.com/MetaMask/metamask-extension/pull/14495/files. I only today realized that `flat` is also browser incompatible. See description above for relevant browser compatibility links.
